client/allocrunner/taskrunner/artifact_hook.go

```diff
@@ -25,24 +26,18 @@ func newArtifactHook(e ti.EventEmitter, logger log.Logger) *artifactHook {
 	return h
 }
 
-func (*artifactHook) Name() string {
-	// Copied in client/state when upgrading from <0.9 schemas, so if you
-	// change it here you also must change it there.
-	return "artifacts"
-}
-
-func (h *artifactHook) Prestart(ctx context.Context, req *interfaces.TaskPrestartRequest, resp *interfaces.TaskPrestartResponse) error {
-	if len(req.Task.Artifacts) == 0 {
-		resp.Done = true
-		return nil
+func (h *artifactHook) createWorkers(req *interfaces.TaskPrestartRequest, resp *interfaces.TaskPrestartResponse, noOfWorkers int, jobsChannel chan *structs.TaskArtifact, errorChannel chan error) {
+	var wg sync.WaitGroup
+	for i := 0; i < noOfWorkers; i++ {
+		wg.Add(1)
+		go h.doWork(req, resp, jobsChannel, errorChannel, &wg)
 	}
+	wg.Wait()
+	close(errorChannel)
+}
 
-	// Initialize hook state to store download progress
-	resp.State = make(map[string]string, len(req.Task.Artifacts))
-
-	h.eventEmitter.EmitEvent(structs.NewTaskEvent(structs.TaskDownloadingArtifacts))
-
-	for _, artifact := range req.Task.Artifacts {
+func (h *artifactHook) doWork(req *interfaces.TaskPrestartRequest, resp *interfaces.TaskPrestartResponse, jobs chan *structs.TaskArtifact, errorChannel chan error, wg *sync.WaitGroup) {
+	for artifact := range jobs {
```

Comment:
```suggestion
        defer wg.Done()
	for artifact := range jobs {
```

Reply:
This should help us avoid any unforeseen goroutine leaks.

Reply:
makes sense. resolved via ec98191